keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+      # that is being checked for a match.
+      selector: |-
+        versions["runtimeVersion"] >= "v12.0.0" && quantities["memory"] >= "8Gi"
+    - name: selected-gpu-1
+      # Same as above in this example, but could also be different.
+      selector: |-
+        versions["runtimeVersion"] >= "v12.0.0" && quantities["memory"] >= "8Gi"
+```
+
+The semantic is that the selector expression must evaluate to true for a
+particular sub-partition to be usable. The initial implementation will use a
+very simplistic approach for satisfying requests: it iterates over the requests
+in order and uses the first sub-partition in a depth-first search which
+matches. This locks that instance into a certain partitioning, which then
+influences whether the next request still can be satisfied. With that in mind,
+a better way to list sub-partitions would be to list the small ones first in
+the NodeResourceSlice, to avoid using a whole GPU for a claim that could have
+used a smaller sub-partition instead.
+
+Eventually, this implementation will have to become smarter. It needs to
+support backtracking and scoring of possible solutions to find the best one
+which satisfies the requests. This could be extended to support finding the
+best solution for a set of pods and their claims.
+
+Another future extension is to express constraints that must be satisfied for
+the selected instances. For example, selecting two cards which are on the same
+PCI root complex may be needed to get the required performance.
+
+Instead of defining a vendor CRD, DRA driver authors or administrators may also
+decide to allow users to create and reference ResourceClaimParameters directly
+in their ResourceClaim. Then the translation step from vendor CRD is not
+needed. The downside is a lack of validation of these user-created
+ResourceClaimParameters.
+
+Resource class parameters are supported the same way. To ensure that
+permissions can be limited to administrators, there's a separate cluster-scoped
+ResourceClassParameters type. Instead of individual requests, one additional
+selector can be specified there which then also must be true for all individual
+requests made with that class:
+
+```yaml
+kind: ResourceClassParameters
+apiVersion: resource.k8s.io/v1alpha2
+
+metadata:
+  name: someArbitraryName
+
+generatedFrom:
+  name: gpu-parameters
+  kind: CardClassParameters
+  apiGroup: dra.example.com
+  uid: foobar-uid
+
+vendorParameters:
+  ...
+
+filters:
+- driverName: cards.dra.example.com
+  partitionable:
+    selector: |-
+      quantities["memory"] <= "16Gi"
+```
+
+In this example, the additional selector expression limits users of this class
+to smaller sub-partitions. Together with limiting the number of claims that
+users are allowed to create for this class (see resource quotas in the core
+KEP) this can ensure that users do not consume too much resources. Allowing
+resource quotas that are based on resource attributes may be a useful future
```

Comment:
Quota is an interesting question. Creating quotas on ResourceClaims doesn't really quite capture it. I can see that we can so something with it via the class as you have described here.

If we had a more concrete representation of the underlying resource (as a KRM object not just a resource claim), you could possibly use standard quota on that. Something to think about for later releases.

Reply:
We were planning on building custom quota controllers to be deployed as part of our driver deployment. Ones that understand our claim parameter objects and can enforce quotas based on their contents. We'd much rather reuse existing (or future ones) if we can figure out how to make one of them work in a generalized way for DRA.

Reply:
I agree, worth thinking about. But not now :grin: 

The core DRA quota PR is still pending: https://github.com/kubernetes/kubernetes/pull/120611. I've put it on hold for further deliberations.